orangecontrib/geo/widgets/owchoropleth.py

```diff
@@ -104,301 +524,480 @@ class Outputs:
+                                         valid_types=(ContinuousVariable,))
+        gui.comboBox(lat_lon_box, self, 'attr_lon', label='Longitude:',
+                     callback=self.setup_plot, model=self.lat_lon_model,
+                     **options)
+
+        gui.comboBox(lat_lon_box, self, 'attr_lat', label='Latitude:',
+                     callback=self.setup_plot, model=self.lat_lon_model,
+                     **options)
+
+        agg_box = gui.vBox(self.controlArea, True)
+        gui.comboBox(agg_box, self, 'agg_func', label='Agg.:',
+                     items=AGG_FUNCS, callback=self.setup_plot, **options)
+
+        self.agg_attr_model = DomainModel(valid_types=(ContinuousVariable,
+                                                       DiscreteVariable))
+        self.agg_attr_combo = gui.comboBox(agg_box, self, 'agg_attr',
+                                           label='Attribute:',
+                                           callback=self.setup_plot,
+                                           model=self.agg_attr_model,
+                                           **options)
+
+        a_slider = gui.hSlider(agg_box, self, 'admin_level', minValue=0,
+                               maxValue=2, step=1, label='Detail:',
+                               createLabel=False, callback=self.setup_plot)
+        a_slider.setFixedWidth(176)
+
+        visualization_box = gui.vBox(self.controlArea, True)
+        b_slider = gui.hSlider(visualization_box, self, "binning_index",
+                               label="Bin width:", minValue=0,
+                               maxValue=max(1, len(self.binnings) - 1),
+                               createLabel=False, callback=self.colors_changed)
+        b_slider.setFixedWidth(176)
+
+        av_slider = gui.hSlider(visualization_box, self, "graph.alpha_value",
+                                minValue=0, maxValue=255, step=10,
+                                label="Opacity:", createLabel=False,
+                                callback=self.colors_changed)
+        av_slider.setFixedWidth(176)
+
+        gui.checkBox(visualization_box, self, "graph.show_legend",
+                     "Show legend",
+                     callback=self.graph.update_legend_visibility)
+
+        control_area_stretch = gui.widgetBox(self.controlArea)
+        control_area_stretch.layout().addStretch(100)
+
+        plot_gui = OWPlotGUI(self)
+        plot_gui.box_zoom_select(self.controlArea)
+        gui.auto_send(self.controlArea, self, "auto_commit")
+
+    @property
+    def effective_variables(self):
+        return [self.attr_lat, self.attr_lon] \
+            if self.attr_lat and self.attr_lon else []
+
+    @property
+    def effective_data(self):
+        return self.data.transform(Domain(self.effective_variables,
+                                          self.data.domain.class_vars,
+                                          self.data.domain.metas))
```

Comment:
Because it is implemented in an openclass widget and keeping classes and metas seemed the most generic solution. In some cases the projection method needs a class variable (freeviz). If not agreed, any child widget can override the properties.

I'm not sure that was correctly thought about, but since the widget is inherited in many places already, I'd leave it as it is (and fix the OWScatterplot, which obviously has a 'duplicated names' bug).

As far as the `owchoropleth` goes: I'd remove the class_vars and metas from the effective_data, because it is never used.